src/host.rs

```diff
@@ -127,19 +127,25 @@ impl Host {
 
     pub(crate) fn accept(&mut self) -> Option<Envelope> {
         let now = Instant::now();
-        let deque = &mut self.listener.as_mut()?.deque;
+        let deque = &self.listener.as_ref()?.deque;
 
-        match deque.front() {
-            Some(Envelope {
-                instructions: DeliveryInstructions::DeliverAt(time),
-                ..
-            }) if *time <= now => {
-                let ret = deque.pop_front();
-                self.bump_version();
-                ret
+        // Iterate in order, skipping "held" envelopes, which is necessary to
+        // avoid front-of-line blocking.
+        for (index, envelope) in deque.iter().enumerate() {
+            match envelope {
+                Envelope {
+                    instructions: DeliveryInstructions::DeliverAt(time),
+                    ..
+                } if *time <= now => {
+                    self.bump_version();
+                    let deque_mut = &mut self.listener.as_mut()?.deque;
+                    return deque_mut.remove(index);
```

Comment:
It's wild that this compiles. At least two things are happening:

1. The borrow on L130 is being released, so as to allow a mut ref.
2. And this is safe because L142 returns and so the immu borrow is never taken again.

Really awesome stuff.

Reply:
Right!? I too stared at this in disbelief.